src/bastion-executor/src/thread_manager.rs

```diff
@@ -0,0 +1,406 @@
+            .finish()
+    }
+}
+
+impl DynamicPoolManager {
+    pub fn new(static_threads: usize, runner: Arc<dyn DynamicRunner + Send + Sync>) -> Self {
+        let dynamic_threads = 1.max(num_cpus::get() - static_threads);
+        Self {
+            static_threads,
+            dynamic_threads,
+            parked_threads: ArrayQueue::new(dynamic_threads),
+            runner,
+            last_frequency: AtomicU64::new(0),
+            frequencies: Mutex::new(VecDeque::with_capacity(
+                FREQUENCY_QUEUE_SIZE.saturating_add(1),
+            )),
+        }
+    }
+
+    pub fn increment_frequency(&self) {
+        self.last_frequency.fetch_add(1, Ordering::Acquire);
+    }
+
+    /// Initialize the dynamic pool
+    /// That will be scaled
+    pub fn initialize(&'static self) {
+        // Static thread manager that will always be available
+        trace!("setting up the static thread manager");
+        (0..self.static_threads).for_each(|_| {
+            let clone = Arc::clone(&self.runner);
+            thread::Builder::new()
+                .name("bastion-driver-static".to_string())
+                .spawn(move || {
+                    Self::affinity_pinner();
+                    clone.run_static(THREAD_PARK_TIMEOUT);
+                })
+                .expect("couldn't spawn static thread");
+        });
+
+        // Dynamic thread manager that will allow us to unpark threads
+        // According to the needs
+        trace!("setting up the dynamic thread manager");
+        (0..self.dynamic_threads).for_each(|_| {
+            let clone = Arc::clone(&self.runner);
+            thread::Builder::new()
+                .name("bastion-driver-dynamic".to_string())
+                .spawn(move || {
+                    Self::affinity_pinner();
+                    let parker = || self.park_thread();
+                    clone.run_dynamic(&parker);
+                })
+                .expect("cannot start dynamic thread");
+        });
+
+        // Pool manager to check frequency of task rates
+        // and take action by scaling the pool accordingly.
+        thread::Builder::new()
+            .name("bastion-pool-manager".to_string())
+            .spawn(move || {
+                let poll_interval = Duration::from_millis(SCALER_POLL_INTERVAL);
+                trace!("setting up the pool manager");
+                loop {
+                    self.scale_pool();
+                    thread::park_timeout(poll_interval);
+                }
+            })
+            .expect("thread pool manager cannot be started");
+    }
+
+    /// Provision threads takes a number of threads that need to be made available.
+    /// It will try to unpark threads from the dynamic pool, and spawn more threads if needs be.
+    pub fn provision_threads(&'static self, n: usize) {
+        for i in 0..n {
+            if !self.unpark_thread() {
+                let new_threads = n - i;
```

Comment:
checked_sub here too

Reply:
i is always between 0 and n, is it necessary ?

Reply:
Oh forget about it.